src/middlewares/Cache.ts

```diff
@@ -0,0 +1,155 @@
+import {objectToQueryString} from "../internal/formatting";
+import {FetchRequest, FetchResponse} from "./Fetch";
+
+import {LocalStorageCache} from "../cache/LocalStorageCache";
+import {MemoryCache} from "../cache/MemoryCache";
+import {RedisCache} from "../cache/RedisCache";
+
+import {IMiddleware} from "../Stack";
+
+// Represents nature of cache entry that is persisted into the configured engine.
+export interface ICacheEntry {
+    value: string;
+    expiry: number; // expiry timestamp in milliseconds
+    metadata?: {[index: string]: any};
+}
+
+// Represents cache options for each cache buckets added to the cache middleware
+export interface IFetchRequestCacheOptions {
+    enable: boolean;
+    path: string; // relative path, shall be resolved from FetchRequest.baseUrl
+    maxAge?: number; // cache ttl in seconds
+    key?: string; // optionally a specific key to be used in place of the generated hash
+}
+
+// Represents a signature of cache bucket, each bucket can respond to certain endpoint, have its own cache key and ttl
+export interface ICacheBucket {
+    options: IFetchRequestCacheOptions;
+}
+
+// Represents the storage engine functionality
+export interface ICacheMiddlewareStore {
+    setItem(key: string, value: string, maxAge: number): void;
+    getItem(key: string): string|null;
+    removeItem(key: string): void;
+}
+
+// Represents the cache middleware global configration
+export interface ICacheConfiguration {
+    maxAge?: number;
+    storage?: string;
```

Comment:
storage shouldn't be of string type. if it's a string, we need to look at the doc what the valid strings are.

should be this instead
``storage?: ICacheMiddlewareStore;``